astropy/io/fits/tests/test_core.py

```diff
@@ -643,24 +643,30 @@ def test_open_file_handle(self):
                 pass
 
         # Opening without explicitly specifying binary mode should fail
-        with pytest.raises(ValueError):
-            with open(self.data("test0.fits")) as handle:
-                with fits.open(handle) as _:
-                    pass
+        with (
+            pytest.raises(ValueError),
+            open(self.data("test0.fits")) as handle,
+            fits.open(handle) as _,
+        ):
```

Comment:
So is the nested `with` construct disallowed by PT012 (cf. also https://github.com/astropy/astropy/pull/16158/files#r1530684773)?
I was not familiar with that form of using `with`; while it is kind of neat, it does not make the nested nature very clear and breaks symmetry with the reading operations that are not raising.

Reply:
It is allowed to nest once with one statement inside the inner statement. Once a second nesting statement is added it complains. 

Allowed
```
with pytest.raises(ValueError), open(self.data("test0.fits")) as handle:
    with fits.open(handle) as _:
        pass
```

Not Allowed
```
with pytest.raises(ValueError):
    with open(self.data("test0.fits")) as handle:
        with fits.open(handle) as _:
            pass
```

As for these changes, @eerovaher thought they might be better with a single with statement. I thought it made sense and made the change. https://github.com/astropy/astropy/pull/16158#discussion_r1513087326

Reply:
My suggestion might not have been a good one in this case. When I had a quick look at the tests I saw a bunch of nested `with`-statements and thought that it might be good to rewrite them as a single multi-line `with`-statement (which is new syntax in Python 3.10). However, the tests here are checking that some specific `with`-statements themselves are raising errors and having nested `with`-statements makes that clearer. So now I would recommend rewriting the nested `with`-statements if and only if the error being checked does not come from one of them.

Reply:
Yes, there seem to be some open issues on Ruff handling those cases right, and the [flake8 rule](https://github.com/m-burst/flake8-pytest-style/blob/master/docs/rules/PT012.md) only lists the one-level nesting, not making clear whether this should apply recursively. Whatever flake8 says, the rationale for not missing an exception from the intended statement (here the `fits.open(handle)`) does not apply even for 3 or more context levels, because that statement simply could not be reached and thus not tested if any of the outer `with`s already failed.